subprojects/gdk_rust/gdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,1212 @@
+                            if let Some(policy) = registry_policy {
+                                info!("inserting policy asset {}", &policy);
+                                registry[policy] = json!({"asset_id": &policy, "name": "Liquid Bitcoin", "ticker": "L-BTC"});
+                            }
+
+                            db_for_registry.insert_asset_registry(&registry).unwrap();
+                            db_for_registry.insert_asset_icons(&icons).unwrap();
+                        }
+                        _ => warn!("Registry or icons are not json"),
+                    }
+                } else {
+                    warn!("Cannot download registry and icons");
+                }
+            }))
+        } else {
+            None
+        };
+
+        let checker = match self.network.id() {
+            NetworkId::Bitcoin(network) => {
+                let mut path: PathBuf = self.data_root.as_str().into();
+                path.push(format!("headers_chain_{}", network));
+                ChainOrVerifier::Chain(HeadersChain::new(path, network)?)
+            }
+            NetworkId::Elements(network) => {
+                let verifier = Verifier::new(network);
+                ChainOrVerifier::Verifier(verifier)
+            }
+        };
+
+        let mut headers_kind = match &self.url {
+            ElectrumUrl::Tls(url, validate) => {
+                let client = electrum_client::Client::new_ssl(url.as_str(), *validate)?;
+                HeadersKind::Tls(
+                    Headers::new(db.clone(), checker, client)?,
+                    url.to_string(),
+                    *validate,
+                )
+            }
+            ElectrumUrl::Plaintext(url) => {
+                let client = electrum_client::Client::new(&url)?;
+                HeadersKind::Plain(Headers::new(db.clone(), checker, client)?, url.to_string())
+            }
+        };
+        let (close_headers, r) = channel();
+        self.closer.senders.push(close_headers);
+        let mut chunk_size = DIFFCHANGE_INTERVAL as usize;
+        let headers_handle = thread::spawn(move || {
+            info!("starting headers thread");
+            'outer: loop {
+                if wait_or_close(&r, sync_interval) {
+                    info!("closing headers thread");
+                    break;
+                }
+
+                loop {
+                    match headers_kind.ask(chunk_size) {
+                        Ok(headers_found) => {
+                            if headers_found == 0 {
+                                chunk_size = 1
+                            } else {
+                                info!("headers found: {}", headers_found);
+                            }
+                        }
+                        Err(Error::InvalidHeaders) => {
+                            // this should handle reorgs and also broke IO writes update
+                            if headers_kind.remove(144).is_err() {
```

Comment:
If a borked/malicious server continuously sends me a bunch of invalid headers, will he eventually drain up the headers index?

Reply:
> If a borked/malicious server continuously sends me a bunch of invalid headers, will he eventually drain up the headers index? 

yes

> What effect will this have on transactions associated with these blocks?

txs up to the old tip will still result validated (because the SPV validation is cached after it's validated once) new txs will not be SPV validated until the chain will be built up again

Reply:
> because the SPV validation is cached after it's validated once

The cache for the SPV validation status should be invalidated following a reorg for the transactions that appeared in the reorged blocks, is it?

Will the number of confirmations shown for reorged-but-previously-spv-validated transactions continue to increase while the headers chain is being rebuilt? I guess probably yes, because the tip height keeps updating independently of the headers chain used for SPV?

How long should it typically take to rebuild the chain? I guess this could take awhile, especially on a mobile internet connection. There might be some attack scenario here where a malicious server drains the headers chain in order to keep the client busy resyncing headers and to disable SPV validation for the resync duration.

I can definitely see how handling reorgs more explicitly adds implementation complexity, but perhaps its worth it? I can give this a shot myself if we decide to pursue that.

Reply:
Consider that at the moment SPV is experimental and disabled so not high priority.
That said the issue you raised makes sense and we have to improve our handling of SPV, I am happy to review your work if you want to give a shot